implement a new CGIR and C code generator

zerbina opened the pull request:
## Summary

Add an all new CGIR, together with a new code generation architecture
using it and a new C code generator.

## Details

### New CGIR

Core design decisions for the language:
* structured control-flow: makes translation to Wasm/`asm.js` easier
* syntax that resembles NimSkull/C: familiar to the someone that
  already knows NimSkull and its AST 
* statically and strongly typed: makes debugging and analysis easier 
* narrow(er) operations: keep decision making out of the code
generators
* built-in simple exception handling

Core design decisions for the IR:
* packed AST for storage; terminals are stored out-of-band and
  are interned
* every node carries source location information, to make high
  precision debug information possible
* one IR for everything: types, statements, expressions

The IR also comes with a grammar and type checker, to help with
debugging, troubleshooting, and codifying the static semantics. It's
always built into the compiler, but due to its overhead, has to be
enabled at run-time by passing `-d:validateCgir` to the compiler.

The old CGIR is still used for the JS and VM code generators, and thus
has to be kept for now.

### New Architecture

The intertwined MIR -> CGIR -> C translation is replaced with
separate MIR -> CGIR and CGIR -> C translation steps. This allows
reusing the MIR -> CGIR parts for other code generators.

Code generator may only support a subset/dialect of the CGIR, which the
MIR -> CGIR facilitates by accepting a set of code generator
capabilities.

As a preparation for incremental compilation, the CGIR has support for
being split multiple units (i.e., modules), though this feature is not
actually used right now.

### Breaking Changes

* features: TLS emulation is not supported anymore. The switch still
  exists, but enabling emulation now causes an error
* features: header generation is not supported anymore
* C FFI: field symbols cannot be used in `emit` statements anymore

### Changes To The Produced C Code

* C scopes reflect the NimSkull-level scopes
* immutable pass-by-reference parameters are marked with `NIM_NOALIAS`
* error handling in `.compilerprocs` is not omitted anymore
* except for returning array values, out parameters are not
  used anymore
* large set operations are implemented as runtime procedures
* updating the `TFrame` instances doesn't use C macros anymore
* non-inline routines are emitted in the order they're processed
  by sem (which is roughly the order they appear in the source code)
* RTTI for nominal types is defined in the nominal type's home module
* RTTI for structural types is defined in the project's entry module
* RTTI is initialized using static initializer expressions

### Tests

* split up `thard_alignment.nim` so that the parts not requiring AVX
  can be run on non-amd64 systems
* change  `thard_alignment.nim`  to use  `-mavx`  instead of 
`-march=native` ,
  so that the test is reproducible

zerbina committed:
[upstream] typemaps: properly hash proc types

zerbina committed:
[upstream] mirtypes: properly handle recursive imported types

Imported object types that contain a pointer to themselves weren't
handled properly, leading to the interior types symbols pointing to the
internal object type, not the imported type.

zerbina committed:
[upstream] sigmatch: use proper container for all varargs

zerbina committed:
[upstream] mirgen: handle views into varargs properly

zerbina committed:
[upstream] unreachable_elim: fix bug

zerbina committed:
[upstream] manual: update the top-level emit documentation

Also adjust the related specification test.

zerbina committed:
[upstream] modulelowering: split top-level emits

zerbina committed:
[upstream] tests: update `tstatic_with_converter`

The section specifier wasn't at the start of the string, meaning it was
placed into the procedure section.

zerbina committed:
[upstream] mirtypes: remove offset tracking

It's not used for anything, nor will it be needed.

zerbina committed:
[upstream] mirtypes: remove tagged union types

Instead of a dedicated tagged union type, tagged union support is now
provided by allowing union fields to be associated with tag fields.

zerbina committed:
[upstream] mirtypes: make the embedded state part of fields

zerbina committed:
[upstream] mirpasses: implement type header init injection

It's misplaced, but injecting the initialization earlier is currently
not possible.

zerbina committed:
utils: fix MIR rendering failing for type header access

There's no field at position -1 in PType-based type representation,
which previously caused rendering to crash.

zerbina committed:
[upstream] mirtypes: implement tagged union branch query

zerbina committed:
[upstream] extccomp: fix `CC` hint not working as intended

zerbina committed:
[upstream/finish] rtchecks: implement bound checks

zerbina committed:
[upstream] vm: fix adding `openArray` to NimNode crashing

zerbina committed:
mirtypes: add `paramType` function

zerbina committed:
[upstream] liftdestructors: use proper nil check

zerbina committed:
[upstream] tailcall_elim: fix callConv lookup

zerbina committed:
[upstream] sem: fix typing with tuples containing views

zerbina committed:
[upstream] lambdalifting: mark env as final

Allows for static cleanup of the environment in some cases.

zerbina committed:
[upstream] mirtypes: add dedicated proc signature types

zerbina committed:
[upstream] proto_mir: consider C aliases for conversions

zerbina committed:
mirtypes: don't use canonical type in `lowerField`

Lookup in generic instance types won't return the expected field ID
otherwise.

zerbina committed:
[upstream] mirgen: use a better shape for `mOffsetOf`

The new shape is a lot more robust and also easier to parse.

zerbina committed:
[upstream] types: consider resolved type classes

zerbina committed:
remove `ndi.nim`

Nim Debug Information files are not generated anymore, making the
module for creating them obsolete.

zerbina committed:
containers: fix parameter types of `[]=`

zerbina committed:
remove the previous C code generator

zerbina committed:
compat: remove unused routines

zerbina committed:
mirgen: remove field name interpolation support

It was always a hack and the new C code generator cannot easily support
it anymore.

zerbina committed:
mirgen: always use cstring operand for emit/asm

zerbina committed:
ast_types: remove obsolete `tfObjHasKids` type flag

zerbina committed:
mirpasses: implement "unscoped" def handling

This was previously taken care of by `cgirgen`, but since that's not
used by the new orchestrator / code generator anymore, the fixup pass
has to happen during the MIR phase.

zerbina committed:
mirgen: don't emit unbounded `mnkToSlice` for empty arrays

Zero-length arrays don't exist at the MIR level (they have length one),
meaning that a MIR type array length query won't yield the expected
results.

zerbina committed:
mirgen: use `mnkArrayConstr` for const openArrays

It's a little more intuitive.

zerbina committed:
types: add a to-do comment for array's pass-by-ref behaviour

Arrays being pass-by-value is now supported by the C code generator.
Since changing `isPassByRef` affects language semantics, it's better to
update it in a separate PR.

zerbina committed:
embedded: implement `nimUnhandledException`

zerbina committed:
mirpasses: turn string constructions into constants

The MIR already has the facilities (i.e., `DataTable`) for interning
construction expressions / literals, so its easier to turn the constant
strings into anonymous constants at the MIR stage than doing it during
MIR -> CGIR translation.

zerbina committed:
mirtrees: add an `items` iterator

zerbina committed:
cboot: implement the entry points

Instead of handling entry point generation in the C code generator
(which is quite convoluted), define them in a `system` module.

This makes them easier to write, to implement, and they also become a
little more portable between backends (important for, e.g., LLVM).

zerbina committed:
backends: insert `Return` after global initialization

Every MIR body needs to end in a terminator.

zerbina committed:
mirtrees: fix `subNodes` ignoring `start` parameter

zerbina committed:
mangling: use the mangled name for imported types

Imported types now use their mangled name rather than the name of the
external type. This doesn't affect how they appear in the generated
code, it just prevents CGIR name collisions when there are multiple
definitions for the same imported type.

zerbina committed:
sets: implement a full `set` runtime

The new code generator isn't going to construct the code for large set
operation itself anymore, instead leaving the work to the runtime.

zerbina committed:
cgir: add a deprecation notice

zerbina committed:
cbackend: decouple IC glue from code generation

Instead of calling directly into the C code generator, the IC glue now
returns a module list, making it backend agnostic.

This is only possible now that C code generation doesn't modify
rodfile contents anymore.

zerbina committed:
add the definitions for the new CGIR

zerbina committed:
implement a new C code generation orchestrator

The bulk of the work is delegated to the (not yet implemented) generic
code generator, with `cbackend` only handling invoking the C code
generator and writing the result to disk.

zerbina committed:
add a new CGIR-based C code generator

zerbina committed:
implement a builder API for constructing CGIR trees

A MIR-style approach to tree construction (i.e., `add` and `subTree`)
doesn't scale, quickly becoming cumbersome. For large parts of the
CGIR, a small macro-based API is going to be used.

zerbina committed:
implement a grammar and type checker for the CGIR

The grammar checker is okay, and quite readable (thanks to the used
macro), but the type checker is quite ugly, partially due to being
crammed into a single module.

zerbina committed:
add a pretty-printer for CGIR AST

zerbina committed:
implement the MIR -> CGIR translation

Given the breadth of the language coming out of the MIR stage, and how
comparatively tiny the CGIR is, the translation is very complex. It's
factored into multiple modules, to keep the complexity at least
somewhat manageable.

zerbina committed:
backends: add `generateNimMain`

zerbina committed:
backends: add `generateThreadTeardownProc`

zerbina committed:
implement a generic backend producing CGIR modules

The generic backend works similar to the previous C backend /
orchestrator but producing CGIR modules instead of C modules/files.

zerbina committed:
main: remove the legacy C backend integration

The new C code generator architecture doesn't need a "backend" object
anymore, always leaving the field in the module graph as nil.

zerbina committed:
cbackend: restore `--genMapping` functionality

zerbina committed:
backends: remove the complex origin module tracking

The new C code generator handles inlining itself, rendering the origin
module tracking obsolete.

zerbina committed:
main: reduce peak memory usage by a little

The VM instance doesn't need to stay around during the mid-end and code
generation phases.

zerbina committed:
tests: disable the `--tlsemulation` tests

The feature is not supported by the C code generator anymore, but it
might be again in the future.

zerbina committed:
tests: fix `tcontrolflow.nim`

The test relied on the compiler producing code that "just works" when
using incorrect `.noreturn` and `.compilerproc` annotation, which the
new code generator does not.

zerbina committed:
tests: make the `tsizeof` test work again

Interpolating field names in `.emit` is not possible anymore, but the
test previously relied on it for using C's `offsetof`. A pointer-
arithmetic-based hack is used instead.

zerbina committed:
memory: mark `nimSetMem` as .compilerproc

`nimSetMem` is called many times at run-time as part of `nimZeroMem`,
and the error checking logic has significant overhead. Marking it as a
.compilerproc causes it to be treated as never raising exceptions nor
panics (which it doesn't).

zerbina committed:
tests: update `tnoalias.nim`

The names used for the `openArray` parameter changed.

zerbina committed:
tests: remove test for https://github.com/nim-lang/nim/issues/7363

The test is both misplaced and codifies behaviour that shouldn't have
been introduced -- it also no longer works with the new code generator.

If the programmer wants the type to be defined, they should make this
explicit to the compiler (e.g., by using an explicit dereference).

zerbina committed:
tests; mark `tunsafeaddr.nim` as failing due to known issue

zerbina committed:
tests: update the expected output for `-d:checkAbi` tests

zerbina committed:
tests: fix `timplicit_move_bug.nim`

Semantic analysis should have rejected the code but currently doesn't.
It causes trouble for the MIR -> CGIR translation and thus had to
be fixed.

zerbina committed:
cboot: fix program parameters being cleared

The globals were not marked as `.noinit`, causing `system` module
initialization to clear them with zero after they have been initialized.

zerbina committed:
system: make `cstringArray` access work again

Importing a pointer-to-array-of-pointers in the way it was done before
no longer works. However, `ptr UncheckedArray[cstring]` is equal to
`char**` at the C level, meaning that it can be used as is.

Exporting the type is necessary for disabling the name linter on
the type.

zerbina committed:
tests: enable `tsplit_into_table.nim`

Thanks to `openArray` not being special-cased by the code generator
anymore, the test now works.

zerbina committed:
cgen: restore support for manual include in `header`

zerbina commented:
Third time's the charm. While far from perfect, I do think the result is a lot better than both the previous attempts *and* the present-day C code generator.

I've spent a lot of time polishing it, but `mir2cg` will have to be rewritten from scratch at some point - it just does too many things.

While I did have to use some macros, I tried to keep them as simple as possible. One major macro is the `build` macro for building CGIR trees, another the `genGrammar` macro, for describing the grammar and generating syntax checks based on it. In both cases, writing the resulting code by hand would be unmaintainable.

Wasm, `asm.js`, and LLVM being usable as target languages was an active consideration while designing the CGIR, and while I haven't tested, the translation from CGIR to the respective langauges *should* be relatively straightforward (except for the FFI bits).

zerbina committed:
mirtypes2cg: use anon array types as pointer targets

This is meant as an accommodation for the C code generator.

zerbina committed:
cgir: remove flexible members and the ptr-to-array type

Both concepts are now represented via zero-length arrays

zerbina committed:
mir2cg: remove leftover debug code

zerbina committed:
mir2cg: use `nimCopyMem` for array assignments

Now that CGIR arrays are translated to C arrays, using normal
assignments for CGIR arrays no longer works.

zerbina committed:
mir2cg: use the `genAsgn` taking an `Expr` where necessary

A proper memcopy can only be omitted when the source expression operand
is a proper `Expr` (otherwise taking the address is not possible), so
the `genAsgn` overload taking an `Expr` value has to be used whenever a
memcopy might be necessary.

zerbina committed:
types: also use pass-by-reference for large sets

They're arrays underneath, which cannot be passed by value in C
(without extra code generator support).

zerbina committed:
cgen: adjust to the CGIR changes & new array handling

* remove the array-in-struct wrapping; only emit array typedefs
* remove the obsolete `cnkPtrToArrayTy` and `cnkFlexField` handling
* translate pointers-to-inline-array to pointers-to-element types;
  taking the address of an array lets the lvalue "decay"
* handle inline array types properly (they only appear in field
  declarations)
* use a common procedure for emitting non-function C declarations

zerbina committed:
cboot: add a missing semicolon

zerbina committed:
mir2cg: return arrays via an out parameter

Not pretty, but it's required now that CGIR arrays are translated
directly to C arrays.

zerbina committed:
cgen: translate nullary procs correctly

The parameter list for functions without any parameters must be `void`
prior to C20.

zerbina committed:
validation: fix `equal` for array types

zerbina committed:
validation: rename `elemType` to `pointeeType`

zerbina commented:
In C, arrays are not first class types: they're implicitly converted to pointers to their first element, array declarations decay to pointers in parameter positions, and returning arrays from a procedure (without a pointer indirection) is not possible.

To keep `mirgen` simpler, I had previously opted to implement them as C arrays wrapped in structs in `cgen` (with `mir2cg` always emitting identified array types, never inlining them). This worked very well, but has one major flaw that I didn't consider: imported types.

```nim
{.emit: "/*TYPESECTION*/ struct Foreign { int x[2]; };".}

type Foreign {.importc: "struct Foreign", nodecl.} =
  x: array[2, cint]
```

In the context of the declaration `x`, `array[2, cint]` actually refers to a *C array*. There's no distinction between NimSkull vs. C arrays in neither the source language nor the MIR, so C arrays can be used where NimSkull arrays are expected, and vice versa.

There are multiple ways to address this problem, but for now, I've simply opted for using the same approach as the previous code generator, namely to translate NimSkull/MIR arrays to C arrays directly and accommodate for the array limitation in the code generator (mostly `mir2cg`). Having pass-by-value arrays would have allowed for some `sink` improvements and the struct wrapping also fixed:

```nim
type Obj = object
  x: ptr array[2, array[2, Obj]]
# the C code generated for `Obj` doesn't compile
```
so this is quite unfortunate.

Still, the changes also allowed for simplifying the CGIR type system a bit, by removing the dedicated pointer-to-array type and flexible struct fields (both are subsumed by zero-length arrays, inspired by LLVM).

zerbina committed:
Merge branch 'devel' into new-code-gen-architecture

zerbina committed:
post merge fix-up

zerbina committed:
mirtypes: use `Continuation` type as `.tailcall` return type

Replacing the return type of .tailcall procedures with the
`Continuation` is tricky to do in `mir2cg`, as it would happen during
type translation, which doesn't have access to a mutable type
environment (and neither should it).

zerbina commented:
I've split out all commits tagged with `[upstream]`, except for https://github.com/nim-works/nimskull/pull/1625/commits/4ff9ec3794c8b330aa147c628ffe113da2ead96d, https://github.com/nim-works/nimskull/pull/1625/commits/e9e50c3d636bacd4d31087433e546f34c088fcae, https://github.com/nim-works/nimskull/pull/1625/commits/4a185e334df79daf9f39b47b6ad5770864126d2c. These changes are either too small, require too many changes of code made obsolete by this PR, or cannot be easily tested/explained.

zerbina committed:
more post-merge fixups

zerbina committed:
backends: run discovery after `bekEmit` events

Emit event handling may cause new entities being registered with the
MIR environment, which too have to be queued for processing.

zerbina committed:
tests: disable `ttlsemulation.nim`

Only parts of the test matrix fail (the ones using `--tlsemulation:on`),
so the test is simply disabled wholesale.

zerbina committed:
mir2cg: update `mShlI` translation

zerbina committed:
cgir: only allow left-shift for uint values

zerbina committed:
mir2cg: fix `mnkToSlice` translation

When the slice length is zero, the array pointer must not be accessed.

zerbina committed:
mirpasses: fix the "move unscoped" pass

An `if` followed by a `scope` can still "contain" defs who are not
"scoped", as it's possible for there to still be defs between the end
of the scope and the end of the `if`.

Aside: using a pseudo basic-block representation instead of a real one
was a major mistake. Oh well.

zerbina committed:
chcks: use `sysAssert` directly in `chckBounds`

`.compilerproc`s are treated as never raising when used in source code,
leading to the necessary error handling being omitted.

zerbina committed:
mir2cg: support constant union constructions

Quite an edge case, with undefined behaviour.

zerbina committed:
mir2cg: add missing ptrcast to procval->closure lowering

zerbina committed:
mir2cg: fix wrong alignment being passed to `alignedDealloc`

zerbina commented:
Some time measurements performed on Windows using

```cmd
hyperfine --warmup 1 "<exe> --compileOnly --verbosity:0 --hints:off --warnings:off compiler/nim.nim"
```

on the compiler sources at [devel](https://github.com/nim-works/nimskull/commit/71315c071027d622415b68669211f12c97972571). The C compiler used is `MinGW GCC 11.1.0`.

| Compiler | Built By | Time Taken |
| - | - | - |
| https://github.com/nim-works/nimskull/pull/1625/commits/3550429d1521189df31c27591dd40173db7f2fdf | [devel](https://github.com/nim-works/nimskull/commit/71315c071027d622415b68669211f12c97972571) |  14.574 s ±  0.143 s |
| [devel](https://github.com/nim-works/nimskull/commit/71315c071027d622415b68669211f12c97972571) | [devel](https://github.com/nim-works/nimskull/commit/71315c071027d622415b68669211f12c97972571) | 15.337 s ±  0.144 s |
| https://github.com/nim-works/nimskull/pull/1625/commits/3550429d1521189df31c27591dd40173db7f2fdf | https://github.com/nim-works/nimskull/pull/1625/commits/3550429d1521189df31c27591dd40173db7f2fdf | 15.240 s ±  0.167 s |

These results show that the new backend is significantly faster at producing C code, but with the produced C code being a lot slower. A surface-level profiling at some earlier point during development only yielded that everything became a little slower, and I haven't taken a deeper look at it yet.

My guess is that that the `.compilerproc` error handling or the out parameter change are the culprit.

saem commented on compiler/backend/backends.nim:
```diff
@@ -311,6 +295,58 @@ proc generateMainProcedure*(graph: ModuleGraph, idgen: IdGenerator,
+  # setup the symbol:
+  result = newSym(skProc, getIdent(graph.cache, "nimTeardownThreadVars"),
+                  nextSymId idgen, owner, unknownLineInfo, {})
+  result.flags.incl sfExportc
```
I'm probably not remembering the semantics correctly, apologies if that's the case.

I thought destructors couldn't raise so can this be marked `sfNeverRaises` or is there something else in there that potentially raises?

zerbina replied:
Yep, this procedure can be marked as `sfNeverRaises`, as it only calls destructors, which cannot raise (not even defects).

```suggestion
  result.flags.incl sfExportc
  result.flags.incl sfNeverRaises
```

saem commented on compiler/backend/cgbackend.nim:
```diff
@@ -0,0 +1,353 @@
+    emits: Emit
+
+  BModuleList = object
+    ## Bundles all global processor state.
```
```suggestion
    ## Bundles all global processor state, acting as an accumulator for
    ## the jumbo module definition.
```

Mostly to borrow the language of the module documentation at the top so there is continuity of language, and hopefully also understanding.

saem commented on compiler/backend/cgbackend.nim:
```diff
@@ -0,0 +1,353 @@
+      ## CGIR translation context for the single output module
+
+  OutModule* = object
+    ## A view into the total CCGIR module, derived from a source module. The
```
```suggestion
    ## A view into the jumbo CCGIR module, derived from a source module. The
```

The alternative to this and the change above would be perhaps to update the module docs at the top to define "total".

zerbina replied:
I think "total" is the clearer/better terminology here.

saem commented on compiler/backend/cgbackend.nim:
```diff
@@ -0,0 +1,353 @@
+
+  # give the init procedure a pre-defined external name
+  for key, m in mlist.modules.pairs:
+    m.init.extname = getInitName(m.sym)
```
You could combine this with the loop above.

saem commented on compiler/backend/cgir.nim:
```diff
@@ -1,3 +1,6 @@
+## **Important:** this module provides the definitions for the *lgeacy* CGIR.
+## It's usage is meant to be phased out, and new code generators must not use.
+##
```
```suggestion
## For the newer version see the `cgir2` module
##
```

saem commented on compiler/backend/cgir2.nim:
```diff
@@ -0,0 +1,349 @@
+    cnkString ## packed interned string value
+    cnkLabel  ## label name (32-bit uint)
+    cnkType   ## reference to a type (by name)
+    cnkDatum  ## reference to an anonymous (by ID)
```
```suggestion
    cnkDatum  ## reference to an anonymous constant (by ID)
```

saem commented on compiler/backend/cgir2.nim:
```diff
@@ -0,0 +1,349 @@
+    ast*: Ast
+      ## all AST for everything that's not types
+    data*: Store[Datum, NodeIndex]
+      ## anonymous constants represented as construction expressions
```
```suggestion
      ## anonymous constants represented as construction expressions
      ## constant datum (id) -> index of corresponding declaration in `ast`
```

I haven't read that far yet, but this is my guess of where it's looking things up.

zerbina replied:
It's the index of the construction expression itself; there's no proper declaration for constants.

Having gained some distance from the code, I think there being no proper declaration syntax (like there is for procedures and globals) is a mistake, as it means that `ast` is not self-contained (ignoring out-of-band literal values), which makes it not possible to write it directly to a file.

zerbina replied:
I've only improved the comment for now.

Just adding a dedicated constant declaration syntax wouldn't make the AST self-contained at this point, as there's still the "loose" storage types of, and I'd rather combine these changes with adding a (somewhat) proper transportation format (plus the associated reader/writer and integration into the compiler thereof).

Also, to maybe clarify, with self-contained I mean that the AST contains all information necessary to process it / work with it (without having to store any of the lookup tables - except the out-of-band values - currently part of `CgModule` alongside it).

saem commented on doc/debug.rst:
```diff
@@ -432,7 +432,8 @@ This is done by using one of the following options with the
   translation to the CGIR
-* `cgir`:option: : shows the CGIR as produced by `cgirgen`
+* `cgir`:option: : shows the CGIR as produced by `mir2cg` (`cgirgen` for the
+  legacy CGIR)
```
```suggestion
* `cgir`:option: : shows the CGIR as produced by `mir2cg` (use `cgirgen` for
  the legacy CGIR)
```

zerbina replied:
What I meant to express here is that when the legacy CGIR is used (which happens when compiling using the VM or JS targets), `--showir:cgir` will show the CGIR produced by `cgirgen`, not the one produced by `mir2cg`.

zerbina replied:
I've reworded the documentation to make it easier to understand (the previous wording was indeed quite confusing).

saem commented on compiler/mir/mirtypes.nim:
```diff
@@ -1249,9 +1267,46 @@ func usizeType*(env: TypeEnv): TypeId {.inline.} =
+# ---- convenience type constructors
+
+func newArray*(env: var TypeEnv, count: Positive, typ: TypeId): TypeId =
+  ## Generates a array type with `count` elements of type `typ`.
```
```suggestion
  ## Generates an array type with `count` elements of type `typ`.
```

saem commented on compiler/backend/cgirbuilder.nim:
```diff
@@ -0,0 +1,260 @@
+  ##
+  ## The following syntax is used for `e`:
+  ##
+  ## * a call expressions describes a tree; they can be arbitrarily nested, and
```
```suggestion
  ## * a call expression describes a tree; it can be arbitrarily nested, and
```

saem commented on compiler/backend/validation.nim:
```diff
@@ -0,0 +1,1501 @@
+## Implements the grammer and type checker for the CGIR's in-memory
```
```suggestion
## Implements the grammar and type checker for the CGIR's in-memory
```

saem commented on compiler/backend/validation.nim:
```diff
@@ -0,0 +1,1501 @@
+    top*: Option[NodeIndex]
+      ## the index of the top-level tree node the error appears in
+    where*: NodeIndex
+      ## the node where the
```
```suggestion
      ## the node where the error occurred
```

saem commented on compiler/backend/rtti.nim:
```diff
@@ -0,0 +1,496 @@
+
+# field positions are hardcoded because looking up the fields to retrieve
+# the position is too cumbersome
+const
```
Not that I would necessarily change it now, but any reason to favour `const` over `enum`?

zerbina replied:
I chose `const` over `enum` because using an `enum` gives the impression that the values having a dedicated type is of significance (at least to me), which is not the case here -- the integer value is what's important here.

Using an `enum` would allow restricting the `addField` templates to just accept these values (and not arbitrary integers), but given that usage of `addField` is confined to just this module, I don't think this would buy much.

saem commented on compiler/mir/rtti_queries.nim:
```diff
@@ -0,0 +1,55 @@
+## Some RTTI-related query routines.
```
Maybe rename this module `rtti_queries`, it's more descriptive.

saem reviewed (approved):
I left a few minor comments, but it looks good to me, nicely done!

zerbina committed:
docs: clarify `cgir` option behaviour

zerbina committed:
cgbackend: improve documentation

zerbina committed:
cgbackend: merge init loops in `generateCode`

zerbina committed:
rtti_helper: rename to `rtti_queries`

zerbina committed:
cgir: improve documentation

zerbina committed:
Apply suggestions from code review

Co-authored-by: Saem Ghani <saemghani+github@gmail.com>

zerbina committed:
Merge branch 'devel' into new-code-gen-architecture

zerbina commented:
Surprisingly, the `thard_alignment.nim` [test failure](https://github.com/nim-works/nimskull/actions/runs/18514336582/job/52762376761) in the Windows CI is gone now. I've only changed comments and renamed various entities in the commits afterwards, meaning that the changes *shouldn't* have caused the problem to go away.

When I last tried to reproduce the failure, I wasn't able to, with both ASAN and UBSAN builds also not reporting any issues. I even tried using the exact same GCC downloaded by the "Install MinGW" step, but that also didn't yield any failures. The line reported by the signal handler, indicates that the error was *during* calling the `q1` closure, meaning that the procedure pointer itself must have been invalid somehow, but even after looking through the generated C code, I don't see how this would have been possible.

What did change between the previous CI run and the most [recent one](https://github.com/nim-works/nimskull/actions/runs/19054419864/job/54422400584) is a new Windows runner image being used ([20251021.76](https://github.com/actions/runner-images/blob/win22/20251021.76/images/windows/Windows2022-Readme.md) compared to [20250929.55](https://github.com/actions/runner-images/blob/win22/20250929.55/images/windows/Windows2022-Readme.md)), with a potentially relevant difference between the two being that the shipped GCC changed (14.2.0 vs. 12.2.0), which could hint at a GCC bug and the workflow accidentally using the wrong GCC.

zerbina commented:
/merge

zerbina commented:
The memory leak is caused by the VM's allocator not freeing the cells it creates upon being destroyed itself. I've split out https://github.com/nim-works/nimskull/pull/1625/commits/fe9115925434708198d9e343fbb909d13124c555 into a separate PR that also implements the missing allocator cleanup (#1653).

zerbina committed:
Merge branch 'devel' into new-code-gen-architecture

zerbina committed:
main: remove outdated `graph.vm` reset

alaviss reviewed (commented):
I don't have the time to look at this PR in detail, but I really liked the shape of the new backend and specifics like how entry points are now written in Skull.

Just wanted to chime in to say that you did a wonderful job :smile:.

zerbina committed:
tests: fix and split up `thard_alignment.nim`

This makes the AVX-using part of the test reproducible (since it's not
using `-march=native` anymore) and allows running the non-AVX-using
part on ARM systems.

zerbina commented:
The `thard_alignment.nim` failure is due to a known bug with the MinGW gcc compiler, where the alignment requirement for over-aligned temporaries isn't met, causing the aligned AVX instruction gcc generates for the `return` of the anonymous procedure returned by `lambdaTest2` to trap.

I've split up the test now, with only the part really requiring the AVX instruction set extension to be enabled to use `-mavx`, which fixes the test failure (the crash was in the non-AVX-requiring part).

zerbina commented:
/merge
